File: kata-setup-java/src/main/java/org/example/ItemBackstagePasses.java
package org.example;

public class ItemBackstagePasses extends Item {

  public ItemBackstagePasses(int sellIn, int quality) {
    super("Backstage passes to a TAFKAL80ETC concert", sellIn, quality);
  }

  @Override
  public void preDecreaseSellIn() {
    if (quality < 50) {
      quality = quality + 1;
    }
    if (sellIn < 11) {
      quality = quality + 1;
    }
    if (sellIn < 6) {
      quality = quality + 1;
    }
  }

  @Override
  public void postDecreaseSellIn() {
    if (sellIn < 0) {
      quality = 0;
    }
  }
}

File: kata-setup-typescript/src/Item.BackstagePasses.ts
import { Item, Itemer } from "./Item.js"

export class BackstagePasses extends Item implements Itemer {
  constructor(sellIn: number, quality: number) {
    super("Backstage passes to a TAFKAL80ETC concert", sellIn, quality)
  }

  updateQuality(newSellIn: number): void {
    if (this.isMaxQuality()) {
      return
    }
    if (newSellIn < 11) {
      this.quality = this.quality + 1
    }
    if (newSellIn < 6) {
      this.quality = this.quality + 1
    }
  }

  updateSellIn(): void {
    this.sellIn--
  }
}
